src/Microsoft.VisualStudio.LanguageServices.Razor/Editor/RazorDirectiveCompletionProvider.cs

```diff
@@ -0,0 +1,185 @@
+
+        [ImportingConstructor]
+        public RazorDirectiveCompletionProvider(RazorCodeDocumentProvider codeDocumentProvider)
+        {
+            if (codeDocumentProvider == null)
+            {
+                throw new ArgumentNullException(nameof(codeDocumentProvider));
+            }
+
+            _codeDocumentProvider = codeDocumentProvider;
+        }
+
+        public override Task<CompletionDescription> GetDescriptionAsync(Document document, CompletionItem item, CancellationToken cancellationToken)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (item == null)
+            {
+                throw new ArgumentNullException(nameof(item));
+            }
+
+            var descriptionContent = new List<TaggedText>();
+            if (item.Properties.TryGetValue(DescriptionKey, out var directiveDescription))
+            {
+                var descriptionText = new TaggedText(TextTags.Text, directiveDescription);
+                descriptionContent.Add(descriptionText);
+            }
+
+            var completionDescription = CompletionDescription.Create(descriptionContent.ToImmutableArray());
+            return Task.FromResult(completionDescription);
+        }
+
+        public override Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (!context.Document.FilePath.EndsWith(".cshtml", StringComparison.OrdinalIgnoreCase))
+            {
+                // Not a Razor file.
+                return Task.FromResult(0);
+            }
+
+            if (!_codeDocumentProvider.TryGetFromDocument(context.Document, out var codeDocument))
+            {
+                // A Razor code document has not yet been associated with the document.
+                return Task.FromResult(0);
+            }
+
+            var syntaxTree = codeDocument.GetSyntaxTree();
+
+            if (!AtDirectiveCompletionPoint(syntaxTree, context))
+            {
+                // Can't have a valid directive at the current location.
+                return Task.FromResult(0);
+            }
+
+            var directives = syntaxTree.Options.Directives.Concat(DefaultDirectives);
+            foreach (var directive in directives)
+            {
+                var propertyDictionary = new Dictionary<string, string>(StringComparer.Ordinal);
+
+                if (!string.IsNullOrEmpty(directive.Description))
+                {
+                    propertyDictionary[DescriptionKey] = directive.Description;
+                }
+
+                var completionItem = CompletionItem.Create(
+                    directive.Directive,
+                    // This groups all Razor directives together
+                    sortText: "_RazorDirective_",
```

Comment:
This text makes it so all the directives show up together in the completion list BUT above other design time variables like `__o`.

Reply:
Do the directive names show up in alphabetical order?

Reply:
si 😄